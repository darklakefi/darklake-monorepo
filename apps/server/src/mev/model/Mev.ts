import { IsEnum, IsIn, IsInt, IsOptional, Max, Min, Validate } from "class-validator";
import { IsValidSolanaAddress } from "../../validator/IsValidSolanaAddress";
import { Type } from "class-transformer";
import { SandwichEvent } from "@prisma/client";
import { formatSolAmount } from "src/utils/blockchain";

export class GetMevTotalExtractedQuery {
  @Validate(IsValidSolanaAddress)
  address: string;
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}

export enum MevAttacksOrderBy {
  "DATE" = "date",
  "AMOUNT_DRAINED" = "amountDrained",
}

export class GetMevAttacksQuery {
  @Validate(IsValidSolanaAddress)
  address: string;

  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(100)
  @Type(() => Number)
  limit?: number;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  offset?: number;

  @IsOptional()
  @IsEnum(MevAttacksOrderBy)
  orderBy?: MevAttacksOrderBy;

  @IsOptional()
  @IsIn(["asc", "desc"])
  direction?: "asc" | "desc";
}

export interface Mev {
  tokenName: string;
  timestamp: number;
  solAmountSent: number;
  solAmountDrained: number;
  transactionHash: string;
}

export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export class MevAttackResponse {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };

  constructor(props: SandwichEvent) {
    this.solAmount = {
      sent: formatSolAmount(props.sol_amount_swap),
      lost: formatSolAmount(props.sol_amount_drained),
    };

    this.tokenName = "UNKNOWN"; // TODO: add token metadata service & get token name
    this.timestamp = +props.occurred_at;
    this.transactions = {
      frontRun: {
        address: "0x0", // TODO: add address
        signature: "0x0",
      },
      victim: {
        address: "0x0",
        signature: props.tx_hash_victim_swap,
      },
      backRun: {
        address: "0x0",
        signature: "0x0",
      },
    };
  }
}
