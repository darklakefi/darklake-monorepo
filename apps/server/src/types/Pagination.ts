export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum PaginatedResponseDataLimit {
  MEV_EVENTS_LIST = 100,
}
