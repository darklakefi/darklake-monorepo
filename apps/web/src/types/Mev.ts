export enum MevAttackSwapType {
  BUY = "BUY",
  SELL = "SELL",
}

export interface MevTransaction {
  address: string;
  signature: string;
}

export interface MevAttack {
  swapType: MevAttackSwapType;
  tokenName: string;
  timestamp: number;
  solAmount: {
    sent: number;
    lost: number;
  };
  transactions: {
    frontRun: MevTransaction;
    victim: MevTransaction;
    backRun: MevTransaction;
  };
}

export interface MevTotalExtracted {
  totalSolExtracted: number;
  totalUsdExtracted?: number;
}

export interface GetMevTotalExtractedResponse {
  data?: MevTotalExtracted;
  processingBlocks: {
    completed: number;
    total: number;
  };
}
