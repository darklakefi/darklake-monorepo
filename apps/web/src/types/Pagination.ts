export type PaginatedResponse<T> = {
  limit: number;
  offset: number;
  total: number;
  result: T;
};

export enum SortDirection {
  ASC = "asc",
  DESC = "desc",
}
