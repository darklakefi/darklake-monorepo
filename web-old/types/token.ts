export interface Token {
  symbol: string;
  name: string;
  image: string;
  address: string;
  tokenProgram: string;
  decimals: number;
}
