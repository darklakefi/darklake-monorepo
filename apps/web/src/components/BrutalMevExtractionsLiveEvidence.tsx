import Link from "next/link";
import MevExtractionCard from "./MevExtractionCard";
import { format } from "date-fns";
import { useId } from "react";
import { MevAttack } from "@/types/Mev";
import { formatPercentage } from "@/utils/number";

interface BrutalMevExtractionsLiveEvidenceProps {
  attacks: MevAttack[];
}

interface MevExtractionProgressBarProps {
  percentageExtracted: number;
}

function MevExtractionProgressBar(props: MevExtractionProgressBarProps) {
  const { percentageExtracted } = props;
  const totalBarCount = 30;

  return (
    <div className="text-lg flex min-w-0 flex-grow-0 overflow-hidden mb-2">
      <div
        className="overflow-hidden"
        style={{
          width: `${percentageExtracted}%`,
        }}
      >
        {"█".repeat(totalBarCount)}
      </div>
      <div
        className="overflow-hidden"
        style={{
          width: `${100 - percentageExtracted}%`,
        }}
      >
        {"░".repeat(totalBarCount)}
      </div>
    </div>
  );
}

export default function BrutalMevExtractionsLiveEvidence(props: BrutalMevExtractionsLiveEvidenceProps) {
  const { attacks } = props;
  const key = useId();

  return (
    <div>
      <h2 className="text-3xl uppercase text-brand-30 font-primary mb-4">
        Brutal MEV Extractions <span className="text-brand-20">Live Evidence</span>
      </h2>
      <div className="flex flex-col md:flex-row gap-4">
        {attacks.map((attack, index) => {
          const { tokenName, timestamp, solAmount } = attack;

          const formattedDate = format(timestamp, "yyyy-MM-dd hh:mm OOO");

          const percentageExtracted = formatPercentage((solAmount.lost / solAmount.sent) * 100);

          return (
            <MevExtractionCard key={`${key}-${index}`}>
              <h3 className="text-lg text-brand-30">Case {index + 1}</h3>
              <span className="text-lg text-brand-30">Token: {tokenName}</span>
              <span className="text-lg text-brand-30 mb-4">Date: {formattedDate}</span>
              <hr className="border-brand-40 mb-4" />
              <span className="text-3xl mb-2">{solAmount.lost} SOL Lost</span>
              <MevExtractionProgressBar percentageExtracted={+percentageExtracted} />
              <span className="text-lg text-brand-20">{percentageExtracted}% Extracted</span>
              <span className="text-lg text-brand-30 mb-4">From a {solAmount.sent} SOL Transaction</span>
              <hr className="border-brand-40 mb-4" />
              <Link href="#" className="text-lg text-brand-30 underline">
                View Attack Breakdown
              </Link>
            </MevExtractionCard>
          );
        })}
      </div>
    </div>
  );
}
