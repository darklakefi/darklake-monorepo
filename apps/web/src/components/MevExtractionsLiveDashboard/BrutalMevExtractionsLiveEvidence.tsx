"use client";

import { useId } from "react";
import { MevAttack } from "@/types/Mev";
import AttackDetailCard from "@/components/AttackDetailCard";
import { useAttackBreakdownModal } from "@/providers/AttackBreakdownModalProvider";

interface BrutalMevExtractionsLiveEvidenceProps {
  attacks: MevAttack[];
}

export default function BrutalMevExtractionsLiveEvidence(props: BrutalMevExtractionsLiveEvidenceProps) {
  const { attacks } = props;
  const key = useId();
  const { openModal } = useAttackBreakdownModal();

  return (
    <div>
      <h2 className="text-3xl uppercase text-brand-30 font-primary mb-4">
        Brutal MEV Extractions <span className="text-brand-20">Live Evidence</span>
      </h2>
      <div className="flex flex-col md:flex-row gap-4">
        {attacks?.map((attack, index) => (
          <div className="flex-1 overflow-hidden" key={`${key}-${index}`}>
            <AttackDetailCard
              cardTitle={<span className="uppercase">Case #{index + 1}</span>}
              mevAttack={attack}
              onOpenModal={() => openModal(attack)}
            />
          </div>
        ))}
      </div>
    </div>
  );
}
