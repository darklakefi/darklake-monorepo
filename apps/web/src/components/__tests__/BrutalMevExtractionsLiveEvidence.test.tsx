import { act, render, RenderResult } from "@testing-library/react";
import BrutalMevExtractionsLiveEvidence from "@/components/BrutalMevExtractionsLiveEvidence";
import { MevAttack, MevAttackSwapType, MevTransaction } from "@/types/Mev";

const mockMevAttackTransaction: MevTransaction = {
  address: "someAddress",
  signature: "someSignature",
};

const mockMevAttacks: MevAttack[] = [
  {
    swapType: MevAttackSwapType.BUY,
    timestamp: +new Date("2024-03-26"),
    tokenName: "TRUMP",
    solAmount: {
      sent: 100,
      lost: 50,
    },
    transactions: {
      frontRun: mockMevAttackTransaction,
      victim: mockMevAttackTransaction,
      backRun: mockMevAttackTransaction,
    },
  },
  {
    swapType: MevAttackSwapType.BUY,
    tokenName: "TRUMP",
    timestamp: +new Date("2024-03-28"),
    solAmount: {
      sent: 200,
      lost: 10,
    },
    transactions: {
      frontRun: mockMevAttackTransaction,
      victim: mockMevAttackTransaction,
      backRun: mockMevAttackTransaction,
    },
  },
];

describe("<BrutalMevExtractionsLiveEvidence />", () => {
  it("renders successfully", async () => {
    let baseElement: RenderResult["baseElement"] | undefined;

    await act(async () => {
      ({ baseElement } = render(<BrutalMevExtractionsLiveEvidence attacks={mockMevAttacks} />));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
