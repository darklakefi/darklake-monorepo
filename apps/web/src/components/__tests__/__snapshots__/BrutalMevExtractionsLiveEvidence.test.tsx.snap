// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrutalMevExtractionsLiveEvidence /> renders successfully 1`] = `
<body>
  <div>
    <div>
      <h2
        class="text-3xl uppercase text-brand-30 font-primary mb-4"
      >
        Brutal MEV Extractions 
        <span
          class="text-brand-20"
        >
          Live Evidence
        </span>
      </h2>
      <div
        class="flex flex-col md:flex-row gap-4"
      >
        <div
          class="flex-1 flex overflow-hidden flex-col p-4 uppercase leading-none bg-brand-60"
        >
          <h3
            class="text-lg text-brand-30"
          >
            Case 
            1
          </h3>
          <span
            class="text-lg text-brand-30"
          >
            Token: 
            TRUMP
          </span>
          <span
            class="text-lg text-brand-30 mb-4"
          >
            Date: 
            2024-03-26 12:00 GMT+0
          </span>
          <hr
            class="border-brand-40 mb-4"
          />
          <span
            class="text-3xl mb-2"
          >
            50
             SOL Lost
          </span>
          <div
            class="text-lg flex min-w-0 flex-grow-0 overflow-hidden mb-2"
          >
            <div
              class="overflow-hidden"
              style="width: 50%;"
            >
              ██████████████████████████████
            </div>
            <div
              class="overflow-hidden"
              style="width: 50%;"
            >
              ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
            </div>
          </div>
          <span
            class="text-lg text-brand-20"
          >
            50
            % Extracted
          </span>
          <span
            class="text-lg text-brand-30 mb-4"
          >
            From a 
            100
             SOL Transaction
          </span>
          <hr
            class="border-brand-40 mb-4"
          />
          <a
            class="text-lg text-brand-30 underline"
            href="#"
          >
            View Attack Breakdown
          </a>
        </div>
        <div
          class="flex-1 flex overflow-hidden flex-col p-4 uppercase leading-none bg-brand-60"
        >
          <h3
            class="text-lg text-brand-30"
          >
            Case 
            2
          </h3>
          <span
            class="text-lg text-brand-30"
          >
            Token: 
            TRUMP
          </span>
          <span
            class="text-lg text-brand-30 mb-4"
          >
            Date: 
            2024-03-28 12:00 GMT+0
          </span>
          <hr
            class="border-brand-40 mb-4"
          />
          <span
            class="text-3xl mb-2"
          >
            10
             SOL Lost
          </span>
          <div
            class="text-lg flex min-w-0 flex-grow-0 overflow-hidden mb-2"
          >
            <div
              class="overflow-hidden"
              style="width: 5%;"
            >
              ██████████████████████████████
            </div>
            <div
              class="overflow-hidden"
              style="width: 95%;"
            >
              ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
            </div>
          </div>
          <span
            class="text-lg text-brand-20"
          >
            5
            % Extracted
          </span>
          <span
            class="text-lg text-brand-30 mb-4"
          >
            From a 
            200
             SOL Transaction
          </span>
          <hr
            class="border-brand-40 mb-4"
          />
          <a
            class="text-lg text-brand-30 underline"
            href="#"
          >
            View Attack Breakdown
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
